### PR TITLE
Scrabble - Anna Elisabeth and Tehut Getahun - Queues

### PR DESCRIPTION
**Pull request by annaewolf**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	|  pretty good! 	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  because we don't need an instance of score, we need a total score by player  |
|   Describe an example of a test you wrote for a _nominal case_. |    checking that words of seven letters get an extra 50 points	|
|   Describe an example of a test you wrote for an _edge case_. |  checking that one letter words would score appropriately 	|
|   What was your final test coverage percentage?	|   100%	|
|   Is there a specific piece of code you'd like feedback on?	|   how to make that draw_Tiles method in TileBag more concise!	|

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+module Scrabble
+  class Scoring
+
+    SCORES = {1 => %w(A E I O U L N R S T), 2 => %w( D G), 3 => %w(B C M P), 4 => %w(F H V W Y), 5 => ["K"], 8=> %w(J X) , 10 => %w(Q Z)}
```
Comment (droberts-sea) [inline lib/scoring.rb:4]: This hash of arrays definitely works, but looking up a letter's score ends up being complicated, involving an `each` loop.

Instead, what if you switched it around and had the keys in this array be letters, and the values be the corresponding score. So it would look like

```ruby
SCORE_CHART = {
  'a': 1,
  'b': 3,
  # ...
}
```

Then looking up a letter would be as simple as `SCORE_CHART[letter]`.

File: lib/scoring.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+          highest_word = [word]
+        elsif self.score(word) == self.score(highest_word[-1])
+          highest_word << word
+        end
+      end
+      highest_word = self.tiebreaker(highest_word) if highest_word[1] != nil
+      highest_word[0]
+    end
+
+    def self.tiebreaker(array)
```
Comment (droberts-sea) [inline lib/scoring.rb:29]: Good work abstracting this out to its own method!

File: lib/scoring.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      winner = ""
+      array.each_with_index do |word, index|
+        if word.length == 7
+          return winner = [word]
+        elsif word.length < array[index - 1].length
+          winner = word
+        else
+          winner = array[0]
```
Comment (droberts-sea) [inline lib/scoring.rb:37]: I'm not sure this logic is correct. If you set winner to be `array[0]`, then you're always assuming the best word is at the front, but that's not always true.

As an example, try `Scrabble::Scoring.tiebreaker(['aaaaa', 'k', 'iiiii'])`, which should give `'k'` but instead gives `'aaaaa'`.

File: lib/player.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    attr_reader :name, :plays, :tiles
+    def initialize(name)
+      @name = name
+      @plays = []
+      @tiles = []
+      @tile_bag = TileBag.new
+    end
+
+    def won?
+      false unless total_score > 100
```
Comment (droberts-sea) [inline lib/player.rb:15]: Does this actually work? It looks like it will return either `false` or `nil`, which are both falsey.

File: specs/player_spec.rb [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        it "is an array upon instantiation" do
+          # skip
+          tofu = Scrabble::Player.new("tofu")
+          tofu.plays.must_be_kind_of Array
+        end
+      end
+
+
+    describe "Test the .play Method" do
+      describe "Test play method returns an integer" do
```
Comment (droberts-sea) [inline specs/player_spec.rb:38]: In general, you don't need to prefix your test names with "Test" - we already know that you're testing something. Try and make it short and sweet - what are you `describe`-ing?

File: lib/tile_bag.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
+module Scrabble
+  class TileBag
+
+    def initialize
+      @bag = bag
+    end
+
+    def bag
+      array = []
+      9.times { array << "A"}
```
Comment (droberts-sea) [inline lib/tile_bag.rb:10]: RE: your cleanup question

All of these lines could read like this:

```ruby
array += ["A"] * 9
```

which might be a little less clunky. However, you're storing a lot of data here and at some point it's just going to be a little cumbersome.

File: lib/tile_bag.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+    def draw_tiles(num)
+      tile = ""
+      player_hand = []
+      num.times do
+        tile = @bag.sample
+        player_hand << tile
+        @bag.each_with_index do |letter, index|
+          if letter == tile
+            @bag.delete_at(index)
```
Comment (droberts-sea) [inline lib/tile_bag.rb:47]: You asked about cleaning up this code - `Array.shuffle` and `Array.pop` might be of interest to you here.

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|   yes	|
|   Both Teammates contributed to the codebase	|   yes	|
|   Regular Commits with meaningful commit messages 	|   yes	|
|  **Scoring class**	|   	|
|  `score` method	|   yes	|
|   Uses appropriate data structure to store the letter score	|   see inline comment	|
|   Appropriately handles edge cases	|   I would like to see some explicit code to handle error cases here. What happens `score` is called on an empty string? What if it's called on `'$%^&'`?	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|   Positive edge cases like the two here are covered, but if you were to implement more error handling you would also need to test it.	|
|  `highest_scoring_word` method	|   logic is mostly there - see inline comment	|
|   Appropriately handles edge cases	|   yes	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	|   yes (via tiebreaker)	|
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|   yes	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   yes	|
|  Returns `true` or `false` from `won?` method |  nope - see inline comment |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |    I would like to see much more thorough testing for Player in general. For example: what happens if you attempt to play a bogus word (like `'ab$%'`)? Does `win?` correctly report whether the player has won, at 50, 99, 100, 120 points? When a word is played, does it appear in the list of played words? |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|   yes	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |   yes, but see inline comment	|
| `tiles_remaining` method returns a count	|   yes	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   Yes. The only question I'm not seeing answered is what happens when you draw the last tile, or more tiles than are in the bag.	|

Good work overall! Other than a couple logic bugs, code is generally clean and readable, and git habits look good.

I would like to see more defensive programing - make sure you're checking your method inputs, and try to anticipate things that could go wrong. Expanding test coverage is also a big piece of this. See above and inline comments for more.
